src/virtual_fs.rs

```diff
@@ -0,0 +1,166 @@
+use std::path::{Path, PathBuf};
+use std::io::{self, Read};
+
+use crate::data::{native_load_disks, Disks, Metadata, UserDirectories};
+
+/// An abstraction over the host system, allowing the file dialog to be used to browse e.g. in
+/// memory filesystems.
+///
+/// # Examples
+///
+/// ```
+/// use egui_file_dialog::{FileDialog, FileSystem};
+///
+/// struct MyFileSystem;
+///
+/// impl FileSystem for MyFileSystem { /* ... */ }
+///
+/// let dialog = FileDialog::from_filesystem(Arc::new(MyFileSystem));
+///
+/// /* Use the file dialog as usual */
+/// ```
+pub trait FileSystem {
+    /// Queries metadata for the given path
+    fn metadata(&self, path: &Path) -> io::Result<Metadata>;
+
+    /// Returns true if the path exists and is a directory
+    fn is_dir(&self, path: &Path) -> bool;
+
+    /// Returns true if the path exists and is a file
+    fn is_file(&self, path: &Path) -> bool;
+
+    /// Get the children of a directory
+    fn read_dir(&self, path: &Path) -> io::Result<Vec<PathBuf>>;
+
+    /// Read a short preview of a text file
+    fn load_text_file_preview(&self, path: &Path, max_chars: usize) -> io::Result<String>;
```

Comment:
This is currently only required when the `information_view` feature is enabled

Reply:
Hmm, having different items in the trait depending on features might break things, how about a default impl of load_text_file_preview which throws an error?

Reply:
Mhh yee true. Then let's leave it like that for now and implement a default for the method

Reply:
Added a default method which throws an "Unsupported" error